cpp/src/hash/md5_hash.cu

```diff
@@ -27,23 +27,314 @@
+                                       size_type offset_begin,
+                                       size_type offset_end,
+                                       md5_intermediate_data* hash_state) const
+{
+  for (int i = offset_begin; i < offset_end; i++) {
+    if (!data_col.is_null(i)) {
+      string_view key     = data_col.element<string_view>(i);
+      uint32_t const len  = static_cast<uint32_t>(key.size_bytes());
+      uint8_t const* data = reinterpret_cast<uint8_t const*>(key.data());
+
+      hash_state->message_length += len;
+
+      // 64 bytes are processed in each hash step
+      constexpr int md5_chunk_size = 64;
+      if (hash_state->buffer_length + len < md5_chunk_size) {
+        std::memcpy(hash_state->buffer + hash_state->buffer_length, data, len);
+        hash_state->buffer_length += len;
+      } else {
+        uint32_t copylen = md5_chunk_size - hash_state->buffer_length;
+        std::memcpy(hash_state->buffer + hash_state->buffer_length, data, copylen);
+        md5_hash_step(hash_state);
+
+        while (len > md5_chunk_size + copylen) {
+          std::memcpy(hash_state->buffer, data + copylen, md5_chunk_size);
+          md5_hash_step(hash_state);
+          copylen += md5_chunk_size;
+        }
+
+        std::memcpy(hash_state->buffer, data + copylen, len - copylen);
+        hash_state->buffer_length = len - copylen;
+      }
+    }
+  }
+}
+
+struct MD5Hash {
+  void __device__ finalize(md5_intermediate_data* hash_state, char* result_location) const
+  {
+    auto const full_length = (static_cast<uint64_t>(hash_state->message_length)) << 3;
+    thrust::fill_n(thrust::seq, hash_state->buffer + hash_state->buffer_length, 1, 0x80);
+
+    // 64 bytes are processed in each hash step
+    constexpr int md5_chunk_size = 64;
+    // 8 bytes for the total message length, appended to the end of the last chunk processed
+    constexpr int message_length_size = 8;
+    // 1 byte for the end of the message flag
+    constexpr int end_of_message_size = 1;
+    if (hash_state->buffer_length + message_length_size + end_of_message_size <= md5_chunk_size) {
+      thrust::fill_n(
+        thrust::seq,
+        hash_state->buffer + hash_state->buffer_length + 1,
+        (md5_chunk_size - message_length_size - end_of_message_size - hash_state->buffer_length),
+        0x00);
+    } else {
+      thrust::fill_n(thrust::seq,
+                     hash_state->buffer + hash_state->buffer_length + 1,
+                     (md5_chunk_size - hash_state->buffer_length),
+                     0x00);
+      md5_hash_step(hash_state);
+
+      thrust::fill_n(thrust::seq, hash_state->buffer, md5_chunk_size - message_length_size, 0x00);
+    }
+
+    std::memcpy(hash_state->buffer + md5_chunk_size - message_length_size,
+                reinterpret_cast<uint8_t const*>(&full_length),
+                message_length_size);
+    md5_hash_step(hash_state);
+
+#pragma unroll
```

Comment:
Probably out of scope for a purely refactoring PR like this one, but just curious: do we know the perf impact of this? This is a pretty small loop so the chances that the compiler would automatically unroll seem high but I guess the hint can't hurt.

Reply:
I do not know the performance impact but I agree it's likely to be unrolled automatically. I plan to investigate (and probably remove) this in a later hashing PR when I get into measuring performance.